packages/psss-server/src/connections/EntityConnection.ts

```diff
@@ -3,44 +3,90 @@
  * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
  */
 
-import { PSSS_ENTITY, PSSS_HIERARCHY } from '../constants';
+import { PSSS_HIERARCHY } from '../constants';
 import { ApiConnection } from './ApiConnection';
 
 const { ENTITY_API_URL = 'http://localhost:8050/v1' } = process.env;
 
-type EntityObject = {
+interface Entity {
   id: string;
   code: string;
   name: string;
-};
+  type: string;
+}
 
-interface EntityFilter {
-  type?: string | string[];
+interface EntityOptions {
+  field?: keyof Entity;
+  fields?: (keyof Entity)[];
+  filter?: Partial<Record<keyof Entity, string | string[]>>;
+  includeRootEntity?: boolean;
 }
 
-interface FetchOptions {
-  fields?: string;
+interface RelationOptions {
+  ancestor?: {
+    filter?: { type: string };
+    field?: keyof Entity;
+  };
+  descendant: {
+    filter: { type: string };
+    field?: keyof Entity;
+  };
+  groupBy?: 'ancestor' | 'descendant';
 }
 
-const buildFilter = (fields: EntityFilter) =>
-  Object.entries(fields)
+type ResponseEntity<T extends EntityOptions> = T['field'] extends string
+  ? string
+  : T['fields'] extends Array<unknown>
+  ? Pick<Entity, T['fields'][number]>
+  : Entity;
+
+const getFieldsParam = (fields?: (keyof Entity)[]) => fields && fields.join(',');
+
+const getFilterParam = (filter?: EntityOptions['filter']) =>
+  filter &&
+  Object.entries(filter)
     .map(([key, value]) => `${key}:${value}`)
     .join(';');
 
-export class EntityConnection extends ApiConnection {
-  baseUrl = ENTITY_API_URL;
+const getEntityParams = (options: EntityOptions) => {
+  const { fields, filter, includeRootEntity, ...restOfOptions } = options;
 
-  fetchCountries = async (options: FetchOptions = {}): Promise<EntityObject[]> => {
-    const { fields } = options;
-    const filter = buildFilter({ type: 'country' });
+  return {
+    ...restOfOptions,
+    fields: getFieldsParam(fields),
+    filter: getFilterParam(filter),
+    includeRootEntity: includeRootEntity ? 'true' : undefined,
+  };
+};
 
-    return this.get(`hierarchy/${PSSS_HIERARCHY}/${PSSS_ENTITY}/descendants`, { fields, filter });
+const getRelationParams = (options: RelationOptions) => {
+  const { ancestor = {}, descendant, ...restOfOptions } = options;
+
+  return {
+    ...restOfOptions,
+    ancestor_filter: getFilterParam(ancestor.filter),
+    ancestor_field: ancestor.field,
+    descendant_filter: getFilterParam(descendant.filter),
+    descendant_field: descendant.field,
   };
+};
 
-  fetchSites = async (countryCode: string, options: FetchOptions = {}): Promise<EntityObject[]> => {
-    const { fields } = options;
-    const filter = buildFilter({ type: 'facility' });
+export class EntityConnection extends ApiConnection {
```

Comment:
Good point, I actually had it like this initially (see [EntityConnection.js](https://github.com/beyondessential/tupaia/blob/psss-phase-2/packages/psss-server/src/connections/EntityConnection.ts) in the base branch).

Thoughts:
* Exposing a single `ApiConnection` in each `*-server` package would result in the most DRY approach. On the other hand, it will introduce code package dependencies between multiple packages just to load a single class, which may not be desirable. This may be more of an issue if we eventually deploy microservices as AWS Lambdas
* I ended up in the more generic/abstract interface because I felt that it actually kept things simpler in the specific routes

My suggestion would be: perhaps keeping the generic functionality (and even types) as `private` methods in the `Connection` classes, and expose more specific `public` API for usage by the routes. How do you feel about that? I'll give it a go and see how it looks

Moving forward, we could consider moving all `Connection` code to the `utils` package, or perhaps a dedicated `connections` package, since it's highly likely we'll use them in a similar way across the app